### PR TITLE
Add simple healthz

### PR DESCRIPTION
**Pull request by eriksywu**
Re: #74 
I have a use-case to have the lifecycle of autoscaler pods managed by kubelet via liveness checks

### CODE REVIEW

**Comment by eriksywu** [comment]
I signed it

File: pkg/autoscaler/health.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/golang/glog"
+)
+
+func (s *AutoScaler) startHealthz() {
+	http.HandleFunc("/healthz", s.healthFn)
+	glog.Fatal(http.ListenAndServe(":8080", nil))
+}
+
+func (s *AutoScaler) healthFn(w http.ResponseWriter, req *http.Request) {
+	if req.Header.Get("CheckConfigMap") == "" {
```
Comment (thockin) [inline pkg/autoscaler/health.go]: I don't think healthz should be able to cause "real work" to happen.  If we need an external trigger, we should tackle that independently.

Comment (eriksywu) [reply pkg/autoscaler/health.go]: ah right, it's not idiomatic to do dependency checks via liveness probes. 

We do have a use-case where our we'd like kubelet to restart the autoscaler if it's not able to reach its configmap. I've refactored healthz handler to return the last cached error instead if specified by the `CheckCachedError` header. Would this be an acceptable pattern?

File: pkg/autoscaler/autoscaler_server.go [hunk 2]
```diff
@@ -40,6 +40,7 @@ type AutoScaler struct {
 	clock         clock.Clock
 	stopCh        chan struct{}
 	readyCh       chan<- struct{} // For testing.
+	health        *healthInfo
```
Comment (eriksywu) [inline pkg/autoscaler/autoscaler_server.go]: realizing this isn't the best naming for what it's doing. will rename after i get some feedback

File: pkg/autoscaler/health.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/golang/glog"
+)
+
+type healthInfo struct {
+	m   *sync.RWMutex
```
Comment (thockin) [inline pkg/autoscaler/health.go]: Unless we expect heavy concurrency of readers, RWMutex is overkill and cognitive burned.  Simple Mutex should suffice.

File: pkg/autoscaler/health.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+func (h *healthInfo) setError(err error) {
+	h.m.Lock()
+	h.err = err
+	h.m.Unlock()
+}
+
+func (h *healthInfo) getError() error {
+	h.m.RLock()
+	defer h.m.RUnlock()
```
Comment (thockin) [inline pkg/autoscaler/health.go]: here you use defer.  Above you use explicit unlock.  Prefer defer forconsistency?

File: pkg/autoscaler/health.go (hunk 3 above)
Comment (thockin) [reply pkg/autoscaler/health.go]: And why is this a pointer?  It should not need to be - copying this struct by value would be wrong.

File: pkg/autoscaler/health.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	return h.err
+}
+
+func (s *AutoScaler) startHealthz() {
+	http.HandleFunc("/healthz", s.healthFn)
+	glog.Fatal(http.ListenAndServe(":8080", nil))
+}
+
+func (s *AutoScaler) healthFn(w http.ResponseWriter, req *http.Request) {
+	if req.Header.Get("CheckCachedError") != "true" {
```
Comment (thockin) [inline pkg/autoscaler/health.go]: I think this is overkill.  

If you want the status of the last poll, OK. Let's expose that.  We don't need a header to decide that.  Just use a different URL path.

/healthz -> is it alive?
/last-poll -> did it sync on the last cycle?

How you use that info is what is interesting.  If the poll cycle and the probe cycle are not configured well, you could end up failing the probe "repeatedly" without having actually polled the server in between.

An alternative approach could be to take a flag indicating how many failures to allow and then exit().

File: pkg/autoscaler/health.go (hunk 3 above)
Comment (eriksywu) [reply pkg/autoscaler/health.go]: ok right, fixing

File: pkg/autoscaler/health.go (hunk 4 above)
Comment (eriksywu) [reply pkg/autoscaler/health.go]: using defer

File: cmd/cluster-proportional-autoscaler/options/options.go [hunk 6]
```diff
@@ -123,4 +124,5 @@ func (c *AutoScalerConfig) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&c.PrintVer, "version", c.PrintVer, "Print the version and exit.")
 	fs.Var(&c.DefaultParams, "default-params", "Default parameters(JSON format) for auto-scaling. Will create/re-create a ConfigMap with this default params if ConfigMap is not present.")
 	fs.StringVar(&c.NodeLabels, "nodelabels", c.NodeLabels, "NodeLabels for filtering search of nodes and its cpus by LabelSelectors. Input format is a comma separated list of keyN=valueN LabelSelectors. Usage example: --nodelabels=label1=value1,label2=value2.")
+	fs.IntVar(&c.MaxAttempts, "max-attempts", c.MaxAttempts, "Number of consecutive polling failures before exiting. Default value of 0 will allow for unlimited retries.")
```
Comment (thockin) [inline cmd/cluster-proportional-autoscaler/options/options.go]: Sorry to nit-pick, can we call this "max-failures" or "max-sync-failures"

File: pkg/autoscaler/autoscaler_server.go [hunk 7]
```diff
@@ -66,25 +77,35 @@ func (s *AutoScaler) Run() {
...
 		}
 	}
 }
 
-func (s *AutoScaler) pollAPIServer() {
+func (s *AutoScaler) tryPollAPIServer() {
+	err := s.pollAPIServer()
+	s.lastPollCycleHealth.setLastPollError(err)
+	// if we've tried polling the apiserver more times than allowed
+	if s.maxAttempts > 0 && s.lastPollCycleHealth.failedCount == s.maxAttempts {
```
Comment (thockin) [inline pkg/autoscaler/autoscaler_server.go]: Might be a bit cleaner to have setLastError() return the new fail count.

File: pkg/autoscaler/health.go (hunk 5 above)
Comment (eriksywu) [reply pkg/autoscaler/health.go]: Thanks for the suggestions!

What do you think about keeping the two endpoints as well as having your suggested alternative (flag to force restart after maxAttempts)? I don't believe they will step over each other. I've pushed code to include both. 

Also did a bit of a refactor to have the health server as a dependency.

File: cmd/cluster-proportional-autoscaler/options/options.go [hunk 8]
```diff
@@ -123,4 +124,5 @@ func (c *AutoScalerConfig) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&c.PrintVer, "version", c.PrintVer, "Print the version and exit.")
 	fs.Var(&c.DefaultParams, "default-params", "Default parameters(JSON format) for auto-scaling. Will create/re-create a ConfigMap with this default params if ConfigMap is not present.")
 	fs.StringVar(&c.NodeLabels, "nodelabels", c.NodeLabels, "NodeLabels for filtering search of nodes and its cpus by LabelSelectors. Input format is a comma separated list of keyN=valueN LabelSelectors. Usage example: --nodelabels=label1=value1,label2=value2.")
+	fs.IntVar(&c.MaxSyncFailures, "max-synce-failures", c.MaxSyncFailures, "Number of consecutive polling failures before exiting. Default value of 0 will allow for unlimited retries.")
```
[the remark this answers is not in the document]
Comment (eriksywu) [reply cmd/cluster-proportional-autoscaler/options/options.go]: thanks

**Comment by eriksywu** [comment]
@thockin can you take another look? thnx

**Comment by eriksywu** [comment]
/assign @MrHohn 
hoping to get this merged soon

**Comment by MrHohn** [comment]
@eriksywu Could you squash the fix-up commits? Thanks.

**Comment by eriksywu** [comment]
Thanks! @MrHohn - is it possible to make a new release on gcr soon?

**Pull request merged**